File: Services/TeamService.js
import { Team } from '../Models/Team.js';
import { Participant } from '../Models/Participant.js';
import { AuditLog } from '../Models/AuditLog.js';

export class TeamService {
  /**
   * Retrieves all teams.
   */
  static async getAllTeams(eventId = 1) {
    return await Team.find({ eventId }).sort({ createdAt: 1 });
  }

  /**
   * Retrieves a single team by ID.
   */
  static async getTeamById(id) {
    return await Team.findById(id);
  }

  /**
   * Creates a new team manually.
   */
  static async createTeam(name, eventId = 1, members = []) {
    const team = new Team({
      name: name || 'New Team',
      eventId,
      status: 'Draft',
      isLocked: false,
      members,
    });
    const saved = await team.save();

    await AuditLog.create({
      changeType: 'TeamCreated',
      teamId: saved._id,
      actor: 'Organizer',
      details: `Team '${saved.name}' created manually.`,
    });

    return saved;
  }

  /**
   * Moves a participant from one team to another.
   */
  static async moveParticipantBetweenTeams(participantId, fromTeamId, toTeamId, actor = 'Organizer') {
    const fromTeam = await Team.findById(fromTeamId);
    const toTeam = await Team.findById(toTeamId);

    if (!fromTeam || !toTeam) {
      throw new Error('One or both teams not found');
    }

    if (fromTeam.isLocked || toTeam.isLocked) {
      throw new Error('Cannot move participants involving a locked team');
    }

    const memberIndex = fromTeam.members.findIndex(
      m => m.participantId.toString() === participantId.toString()
    );

    if (memberIndex === -1) {
      throw new Error('Participant not found in source team');
    }

    const [movingMember] = fromTeam.members.splice(memberIndex, 1);
    await fromTeam.save();

    toTeam.members.push({
      participantId: movingMember.participantId,
      name: movingMember.name,
      email: movingMember.email,
      skills: movingMember.skills || [],
      isLocked: false,
      addedAt: new Date(),
    });
    await toTeam.save();

    // Audit log
    await AuditLog.create({
      changeType: 'ParticipantMoved',
      participantId: movingMember.participantId,
      teamId: toTeam._id,
      actor,
      oldValue: fromTeam.name,
      newValue: toTeam.name,
      details: `Moved '${movingMember.name}' from '${fromTeam.name}' to '${toTeam.name}'`,
    });

    return { fromTeam, toTeam };
  }

  /**
   * Toggles team lock state.
   */
  static async toggleTeamLock(teamId, actor = 'Organizer') {
    const team = await Team.findById(teamId);
    if (!team) {
      throw new Error('Team not found');
    }

    const previousLock = team.isLocked;
    team.isLocked = !previousLock;
    team.status = team.isLocked ? 'Locked' : 'Draft';
    await team.save();

    await AuditLog.create({
      changeType: team.isLocked ? 'TeamLocked' : 'TeamUnlocked',
      teamId: team._id,
      actor,
      oldValue: previousLock ? 'Locked' : 'Draft',
      newValue: team.status,
      details: `Team '${team.name}' ${team.isLocked ? 'locked' : 'unlocked'} by ${actor}`,
    });

    return team;
  }

  /**
   * Locks or unlocks a specific participant within a team.
   */
  static async toggleParticipantLockInTeam(teamId, participantId, actor = 'Organizer') {
    const team = await Team.findById(teamId);
    if (!team) {
      throw new Error('Team not found');
    }

    const member = team.members.find(
      m => m.participantId.toString() === participantId.toString()
    );
    if (!member) {
      throw new Error('Participant not found in team');
    }

    member.isLocked = !member.isLocked;
    await team.save();

    await AuditLog.create({
      changeType: 'ParticipantLocked',
      teamId: team._id,
      participantId: member.participantId,
      actor,
      details: `Participant '${member.name}' in team '${team.name}' lock set to ${member.isLocked}`,
    });

    return team;
  }

  /**
   * Retrieves audit logs.
   */
  static async getAuditLogs(limit = 100) {
    return await AuditLog.find().sort({ timestamp: -1 }).limit(limit);
  }

  /**
   * Computes portal statistics.
   */
  static async getStats() {
    const totalParticipants = await Participant.countDocuments();
    const totalTeams = await Team.countDocuments();
    const teams = await Team.find();
    
    let totalMembersInTeams = 0;
    for (const t of teams) {
      totalMembersInTeams += t.members.length;
    }
    const avgTeamSize = totalTeams > 0 ? (totalMembersInTeams / totalTeams).toFixed(1) : 0;

    return {
      totalParticipants,
      totalTeams,
      availableSkills: '24',
      avgTeamSize,
    };
  }

  /**
   * Exports data to CSV string format.
   */
  static async exportCSV(type) {
    if (type === 'participants') {
      const participants = await Participant.find().sort({ createdAt: -1 });
      let csv = 'ID,Name,Email,Profile Status,Skills,Created At\n';
      for (const p of participants) {
        const skills = (p.skills || []).map(s => `${s.skillName}(${s.proficiencyLevel})`).join('; ');
        csv += `"${p._id}","${p.name}","${p.email}","${p.profileStatus}","${skills}","${p.createdAt.toISOString()}"\n`;
      }
      return csv;
    }

    if (type === 'teams') {
      const teams = await Team.find().sort({ createdAt: 1 });
      let csv = 'Team ID,Team Name,Status,Is Locked,Member Count,Members,Created At\n';
      for (const t of teams) {
        const memberNames = (t.members || []).map(m => m.name).join('; ');
        csv += `"${t._id}","${t.name}","${t.status}",${t.isLocked},${t.members.length},"${memberNames}","${t.createdAt.toISOString()}"\n`;
      }
      return csv;
    }

    if (type === 'audit') {
      const logs = await AuditLog.find().sort({ timestamp: -1 });
      let csv = 'ID,Change Type,Actor,Old Value,New Value,Details,Timestamp\n';
      for (const l of logs) {
        csv += `"${l._id}","${l.changeType}","${l.actor}","${l.oldValue || ''}","${l.newValue || ''}","${l.details || ''}","${l.timestamp.toISOString()}"\n`;
      }
      return csv;
    }

    throw new Error('Invalid export type. Supported: participants, teams, audit');
  }
}

File: Services/TeamService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TeamWeave.Models;

namespace TeamWeave.Services
{
    public class TeamService
    {
        private readonly List<Team> _teams = new List<Team>();
        private readonly List<AuditLog> _auditLogs = new List<AuditLog>();
        private int _nextTeamId = 1;
        private int _nextAuditLogId = 1;

        /// <summary>
        /// Creates a new team.
        /// </summary>
        public Team CreateTeam(string name, int eventId)
        {
            var team = new Team
            {
                Id = _nextTeamId++,
                Name = name,
                EventId = eventId,
                Status = "Draft",
                IsLocked = false
            };

            _teams.Add(team);
            return team;
        }

        /// <summary>
        /// Adds a participant to a team.
        /// </summary>
        public void AddParticipantToTeam(int teamId, Participant participant)
        {
            var team = GetTeamById(teamId);
            if (team == null)
                throw new InvalidOperationException("Team not found");

            if (team.IsLocked)
                throw new InvalidOperationException("Cannot modify locked team");

            var teamMember = new TeamMember
            {
                TeamId = teamId,
                ParticipantId = participant.Id,
                IsLocked = false,
                Participant = participant,
                Team = team
            };

            team.Members.Add(teamMember);
            participant.Teams.Add(teamMember);
            team.UpdatedAt = DateTime.Now;

            // Log the action
            LogAuditEvent("ParticipantAdded", teamId, participant.Id, null, participant.Name, "System");
        }

        /// <summary>
        /// Moves a participant from one team to another.
        /// </summary>
        public void MoveParticipantBetweenTeams(int participantId, int fromTeamId, int toTeamId)
        {
            var fromTeam = GetTeamById(fromTeamId);
            var toTeam = GetTeamById(toTeamId);

            if (fromTeam == null || toTeam == null)
                throw new InvalidOperationException("One or both teams not found");

            if (fromTeam.IsLocked || toTeam.IsLocked)
                throw new InvalidOperationException("Cannot modify locked teams");

            var memberToMove = fromTeam.Members.FirstOrDefault(m => m.ParticipantId == participantId);
            if (memberToMove == null)
                throw new InvalidOperationException("Participant not found in source team");

            // Remove from source team
            fromTeam.Members.Remove(memberToMove);
            fromTeam.UpdatedAt = DateTime.Now;

            // Add to destination team
            var newMember = new TeamMember
            {
                TeamId = toTeamId,
                ParticipantId = participantId,
                IsLocked = false,
                Participant = memberToMove.Participant,
                Team = toTeam
            };

            toTeam.Members.Add(newMember);
            toTeam.UpdatedAt = DateTime.Now;

            // Log the action
            LogAuditEvent("ParticipantMoved", toTeamId, participantId, 
                fromTeam.Name, toTeam.Name, "System");
        }

        /// <summary>
        /// Locks a team to prevent further modifications.
        /// </summary>
        public void LockTeam(int teamId, string actor = "System")
        {
            var team = GetTeamById(teamId);
            if (team == null)
                throw new InvalidOperationException("Team not found");

            team.IsLocked = true;
            team.Status = "Locked";
            team.UpdatedAt = DateTime.Now;

            LogAuditEvent("TeamLocked", teamId, null, "false", "true", actor);
        }

        /// <summary>
        /// Unlocks a team to allow modifications.
        /// </summary>
        public void UnlockTeam(int teamId, string actor = "System")
        {
            var team = GetTeamById(teamId);
            if (team == null)
                throw new InvalidOperationException("Team not found");

            team.IsLocked = false;
            team.Status = "Draft";
            team.UpdatedAt = DateTime.Now;

            LogAuditEvent("TeamUnlocked", teamId, null, "true", "false", actor);
        }

        /// <summary>
        /// Locks a specific participant in a team.
        /// </summary>
        public void LockParticipantInTeam(int teamId, int participantId)
        {
            var team = GetTeamById(teamId);
            if (team == null)
                throw new InvalidOperationException("Team not found");

            var member = team.Members.FirstOrDefault(m => m.ParticipantId == participantId);
            if (member == null)
                throw new InvalidOperationException("Participant not found in team");

            member.IsLocked = true;
            team.UpdatedAt = DateTime.Now;

            LogAuditEvent("ParticipantLocked", teamId, participantId, "false", "true", "System");
        }

        /// <summary>
        /// Retrieves a team by ID.
        /// </summary>
        public Team GetTeamById(int id)
        {
            foreach (var team in _teams)
            {
                if (team.Id == id)
                    return team;
            }
            return null;
        }

        /// <summary>
        /// Retrieves all teams.
        /// </summary>
        public List<Team> GetAllTeams()
        {
            return new List<Team>(_teams);
        }

        /// <summary>
        /// Retrieves all audit logs.
        /// </summary>
        public List<AuditLog> GetAuditLogs()
        {
            return new List<AuditLog>(_auditLogs);
        }

        /// <summary>
        /// Retrieves audit logs for a specific team.
        /// </summary>
        public List<AuditLog> GetAuditLogsForTeam(int teamId)
        {
            return _auditLogs.Where(log => log.TeamId == teamId).ToList();
        }

        private void LogAuditEvent(string changeType, int? teamId, int? participantId, 
            string oldValue, string newValue, string actor)
        {
            var auditLog = new AuditLog
            {
                Id = _nextAuditLogId++,
                ChangeType = changeType,
                TeamId = teamId,
                ParticipantId = participantId ?? 0,
                OldValue = oldValue,
                NewValue = newValue,
                Actor = actor,
                Timestamp = DateTime.Now
            };

            _auditLogs.Add(auditLog);
        }
    }
}
